streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java

```diff
@@ -214,18 +214,13 @@ public long streamTime() {
         return streamTimeSupplier.get();
     }
 
-    private abstract static class StateStoreReadOnlyDecorator<T extends StateStore> extends AbstractStateStore {
+    private abstract static class StateStoreReadOnlyDecorator<T extends StateStore> extends WrappedStateStore<T> {
```

Comment:
This is a meta thought orthogonal to this PR, but I'd leave it here and then create a JIRA ticket later: The `StateStore` interface is really designed to be "implemented by users, and called by library" -- maybe one can argue that users may want to explicitly call `flush` as well, but I'd say it is okay to NOT allow users do so. And the extended interface, like `KeyValueStore` are designed to be "implemented by library or user-customizable, and called by user".

Given this thought, we should probably consider refactoring the hierarchy of state stores more as:

1) `interface StateStore`
2) `interface ReadOnlyKeyValueStore`
3) `interface KeyValueStateStore extends StateStore, ReadOnlyKeyValueStore` (note the name difference)
4) `interface KeyValueStore extends ReadOnlyKeyValueStore`.

And 3) would be the one used in `Stores` factory, i.e. required for either built-in or user-customization, while 4) would be the one returned from `ProcessorContext.getStateStore`. The returned object would, in fact, always be extending `StateStore` as well but only library will care to cast and call its inner functions.

I'm not sure if there's a better way to do this in order to get rid of all such decorators, @vvcephei wdyt?

Reply:
While I completely agree (all the "decorators" are only necessary because we got the API wrong, leaking "internal" API like `flush()`, `close()` etc into PAPI).

However, I this possible to refactor without breaking the public API?

Reply:
I agree 100%, and would very much like to see something like this. It would also help us to avoid leaking other administrative methods if we add them (besides just `init`, `flush`, `close`).

I think cleanest the way to do it backwards compatibly is to introduce a new store hierarchy and deprecate the old one.

Reply:
> I think cleanest the way to do it backwards compatibly is to introduce a new store hierarchy and deprecate the old one.

I am afraid that might be the only way to do it -- deprecation hell -- users won't be happy -- thus we should think hard if it's worth to do or not...

Reply:
Cool, I will create a JIRA ticket for now to keep track of it.